File: questions/0845. Longest Mountain in Array/longest.py
class Solution:
    def longestMountain(self, A: List[int]) -> int:
        l,res = len(A), 0
        down = [0]*l
        up = [0]*l
        for i in range(l - 2, 0, -1):
            if A[i] > A[i + 1]:
                down[i] = down[i + 1] + 1
        for i in range(0,l):
            if i > 0 and A[i] > A[i - 1]:
                up[i] = up[i - 1] + 1
            if down[i] > 0 and up[i] > 0:
                res = max(res, down[i] + up[i] + 1)
        return res

File: questions/159. Longest Substring with At Most Two Distinct Characters/longest.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int left = 0, right = 0, l = s.length(), res = 0;
        HashMap<Character, Integer> map = new HashMap();
        while(right < l){
            map.put(s.charAt(right), map.getOrDefault(s.charAt(right), 0) + 1);
            if(map.size() > 2){
                while(map.size() > 2){
                    map.put(s.charAt(left), map.get(s.charAt(left)) - 1);
                    if(map.get(s.charAt(left)) == 0){
                        map.remove(s.charAt(left));
                    }
                    left++;
                }
            }
            res = Math.max(res, right - left + 1);
            right++;
        }
        return res;
    }
}
